### PR TITLE
feat: Implement Client

### PR DESCRIPTION
**Pull request by ghost**
## Description
Added the implementation of the UIPath client

## Added
 - .gitignore
 - asset.go
 - auth.go
 - client.go
 - client_test.go
 - error.go
 - examples/main.go
 - go.mod
 - go.sum
 - queue_item.go

## Reference
[DGM2-8158](https://comvex.myjetbrains.com/youtrack/issue/DGM2-8158)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,4 @@
+.DS_Store
+.idea/
+.pre-commit-config.yaml
+coverage.out
```
Comment (anderson-comvex) [inline .gitignore]: Let's add the trailing line.

```suggestion
coverage.out

```

File: examples/main.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
+package main
```
Comment (anderson-comvex) [inline examples/main.go]: I don't think the example should be part of the package `main`. It makes hard to exclude from the production build.

File: client_test.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+		},
+	}
+
+	res, err := suite.c.SendWithAuthorization("GET", "/", nil, nil, nil)
+
+	assert.Equal(suite.T(), err, nil)
+	assert.Equal(suite.T(), string(res), resBody)
+}
+
+func (s *ClientTestSuite) PrepareUIPathAuthAPIResponder(providedData OauthTokenResp, envBaseUrl string, endpoint string, method string, HTTPStatusCode int, success bool) {
```
Comment (anderson-comvex) [inline client_test.go]: Rename all suite to `s` or all `s` to `suite`.

```suggestion
func (suite *ClientTestSuite) PrepareUIPathAuthAPIResponder(providedData OauthTokenResp, envBaseUrl string, endpoint string, method string, HTTPStatusCode int, success bool) {
```

File: asset.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+	Value        string `json:"Value"`
+	DisplayValue string `json:"DisplayValue,omitempty"`
+}
+
+// GetByID fetches the asset by id
+func (a *AssetHandler) GetByID(ID uint, folderID uint) (Asset, error) {
+	var asset Asset
+
+	headers := map[string]string{
+		"X-UIPATH-OrganizationUnitId": strconv.Itoa(int(folderID)),
```
Comment (anderson-comvex) [inline asset.go]: Can we move the `"X-UIPATH-OrganizationUnitId"` to a constant?

File: asset.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+	}
+
+	if err = json.Unmarshal(resp, &asset); err != nil {
+		return asset, err
+	}
+
+	return asset, nil
+}
+
+// GetByID fetches the asset by name
```
Comment (anderson-comvex) [inline asset.go]:
```suggestion
// GetByName fetches the asset by name
```

**Review by anderson-comvex — CHANGES_REQUESTED** [review]

File: asset.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+func (a *AssetHandler) GetByName(name string, folderID uint) (Asset, error) {
+	var asset Asset
+
+	headers := map[string]string{
+		"X-UIPATH-OrganizationUnitId": strconv.Itoa(int(folderID)),
+	}
```
Comment (anderson-comvex) [inline asset.go]: This seems to be doing all the same thing every time. Isn't worth to do it in one place in the `a` Asset Handler instance instead?

**Review by anderson-comvex — CHANGES_REQUESTED** [review]

File: client.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+	req.Header.Set("X-UIPATH-TenantName", c.Credentials.TenantName)
+
+	attachHeaders(req, headers)
+
+	resp, err := c.HttpClient.Do(req)
+	if err != nil {
+		return jsonBody, err
+	}
+
+	defer resp.Body.Close()
```
Comment (anderson-comvex) [inline client.go]: Please handle the error in a closure.

```suggestion
	defer func(Body io.ReadCloser) {
		err := Body.Close()
		if err != nil {
			
		}
	}(resp.Body)
```

**Review by anderson-comvex — CHANGES_REQUESTED** [review]

File: error.go [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+)
+
+// RequestError defines how the error looks like from the response
+type RequestError struct {
+	Message          string `json:"message"`
+	ErrorName        string `json:"error"`
+	ErrorCode        int    `json:"errorCode"`
+	ErrorDescription string `json:"error_description"`
+	TraceIDs         string `json:"traceId"`
+	ResoureIds       []uint `json:"resourceIds"`
```
Comment (anderson-comvex) [inline error.go]:
```suggestion
	ResourceIds       []uint `json:"resourceIds"`
```

**Review by anderson-comvex — CHANGES_REQUESTED** [review]

File: queue_item.go [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+	PriorityNormal = "Normal"
+	PriorityHigh   = "High"
+)
+
+// QueueItemHandler struct defines what the asset handler looks like
+type QueueItemHandler struct {
+	Client *Client
+}
+
+// Queue struct defines what the asset model looks like
```
Comment (anderson-comvex) [inline queue_item.go]:
```suggestion
// QueueItem struct defines what the asset model looks like
```

**Review by anderson-comvex — CHANGES_REQUESTED** [review]

File: client.go [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+	respBody, err := ioutil.ReadAll(resp.Body)
+
+	// Handle any errors from the response
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
+		return respBody, ErrorResponseHandler(resp.StatusCode, respBody)
+	}
+
+	return respBody, err
```
Comment (rdehaas-comvex) [inline client.go]: We need to return the `ReadAll` error immediately.
```suggestion
	respBody, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return jsonBody, err
	}

	// Handle any errors from the response
	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
		err = ErrorResponseHandler(resp.StatusCode, respBody)
	}
	
	return respBody, err
```

File: client.go [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+	}
+
+	headersWithAuthorization["Authorization"] = "Bearer " + c.Credentials.Token
+
+	jsonBody, err := json.Marshal(body)
+	if err != nil {
+		return jsonBody, err
+	}
+
+	return c.Send(requestMethod, url, body, headersWithAuthorization, queryParams)
```
Comment (rdehaas-comvex) [inline client.go]: Is it possible to reuse the headers from the parameters?

```suggestion
        headers["Authorization"] = "Bearer " + c.Credentials.Token

	jsonBody, err := json.Marshal(body)
	if err != nil {
		return jsonBody, err
	}

	return c.Send(requestMethod, url, body, headers, queryParams)
```

**Review by rdehaas-comvex — CHANGES_REQUESTED** [review]

File: examples/example.go [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+		Credentials: credentials,
+		URLEndpoint: "{{URLEndpoint}}",
+	}
+
+	res, err := uipath.GetOAuthToken(&c)
+	if err != nil {
+		fmt.Println(res, err)
+	}
+
+	c.Credentials.Token = res.AccessToken
```
Comment (cliff-comvex) [inline examples/example.go]: Can we handle the accessToken generation inside the package? Similar implementation we do in the [digima-backend-client`](https://github.com/comvex-jp/digima-backend-client-go)?

**Review by cliff-comvex — CHANGES_REQUESTED** [review]

**Review by cliff-comvex — CHANGES_REQUESTED** [review]
Can you provide an `example_test.go` on how the mocking should be handle in the client side?

File: queue_item.go [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+	url := fmt.Sprintf("%s%s", q.Client.URLEndpoint, "QueueItems")
+
+	resp, err := q.Client.SendWithAuthorization("GET", url, nil, q.buildHeaders(), filters)
+	if err != nil {
+		return queueItemList.Value, queueItemList.Count, err
+	}
+
+	if err = json.Unmarshal(resp, &queueItemList); err != nil {
+		return queueItemList.Value, queueItemList.Count, err
+	}
```
Comment (anderson-comvex) [inline queue_item.go]: This return value and the below is exactly the same. We dont need this additional `if` here.

```suggestion
    err = json.Unmarshal(resp, &queueItemList)
```

File: queue_item.go [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+	resp, err := q.Client.SendWithAuthorization("GET", url, nil, q.buildHeaders(), map[string]string{})
+	if err != nil {
+		return queueItem, err
+	}
+
+	if err = json.Unmarshal(resp, &queueItem); err != nil {
+		return queueItem, err
+	}
+
+	return queueItem, nil
```
Comment (anderson-comvex) [inline queue_item.go]: We can simplify this error handling.

```suggestion
	err = json.Unmarshal(resp, &queueItem)

	return queueItem, err
```

File: queue_item.go [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+	resp, err := q.Client.SendWithAuthorization("POST", url, queueItemCreateRequest, q.buildHeaders(), map[string]string{})
+	if err != nil {
+		return result, err
+	}
+
+	if err = json.Unmarshal(resp, &result); err != nil {
+		return result, err
+	}
+
+	return result, nil
```
Comment (anderson-comvex) [inline queue_item.go]: Let's simplify the error handling here.

```suggestion
	err = json.Unmarshal(resp, &result)

	return result, err
```

**Review by anderson-comvex — CHANGES_REQUESTED** [review]

File: asset.go [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+	resp, err := a.Client.SendWithAuthorization("GET", url, nil, a.buildHeaders(), map[string]string{})
+	if err != nil {
+		return asset, err
+	}
+
+	if err = json.Unmarshal(resp, &asset); err != nil {
+		return asset, err
+	}
+
+	return asset, nil
```
Comment (anderson-comvex) [inline asset.go]: We can simplify this error handling here.

```suggestion
	err = json.Unmarshal(resp, &asset)

	return asset, err
```

File: asset.go [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
```
Comment (anderson-comvex) [inline asset.go]:
```suggestion
	err = json.Unmarshal(resp, &asset)

	return asset, err
```

File: asset.go [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+	resp, err := a.Client.SendWithAuthorization("GET", url, nil, a.buildHeaders(), filters)
+	if err != nil {
+		return assetList.Value, assetList.Count, err
+	}
+
+	if err = json.Unmarshal(resp, &assetList); err != nil {
+		return assetList.Value, assetList.Count, err
+	}
+
+	return assetList.Value, assetList.Count, nil
```
Comment (anderson-comvex) [inline asset.go]:
```suggestion
	err = json.Unmarshal(resp, &assetList)

	return assetList.Value, assetList.Count, err
```

File: asset.go [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+	resp, err := a.Client.SendWithAuthorization("POST", url, asset, a.buildHeaders(), map[string]string{})
+	if err != nil {
+		return result, err
+	}
+
+	if err = json.Unmarshal(resp, &result); err != nil {
+		return result, err
+	}
+
+	return result, nil
```
Comment (anderson-comvex) [inline asset.go]:
```suggestion
	err = json.Unmarshal(resp, &result)

	return result, err
```

File: asset.go [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+	if err != nil {
+		return asset, err
+	}
+
+	updatedAsset, err := a.GetByID(asset.ID)
+	if err != nil {
+		return asset, err
+	}
+
+	return updatedAsset, nil
```
Comment (anderson-comvex) [inline asset.go]:
```suggestion
	return updatedAsset, err
```

**Review by anderson-comvex — CHANGES_REQUESTED** [review]

File: error.go [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+	}
+
+	return fmt.Sprintf("Request Failed: Error Code(%d) %s", r.ErrorCode, r.Message)
+}
+
+//ErrorResponseHandler handles the errors from the uipath response
+func ErrorResponseHandler(statusCode int, errResp []byte) error {
+	var requestError RequestError
+
+	if len(errResp) < 1 {
```
Comment (anderson-comvex) [inline error.go:33]: How exactly this works? It is kind of not clear to me 🤔  Do you mind adding a comment?

File: client.go [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+	if err != nil {
+		return jsonBody, err
+	}
+
+	defer func(Body io.ReadCloser) {
+		err := Body.Close()
+		if err != nil {
+
+		}
+	}(resp.Body)
```
Comment (anderson-comvex) [inline client.go]: When i added this comment i meant you to actually handle the error 😅 .

See some examples

https://trstringer.com/golang-deferred-function-error-handling/

**Review by anderson-comvex — CHANGES_REQUESTED** [review]

File: auth.go [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+	respBody, err := c.Send("POST", url, body, map[string]string{}, nil)
+	if err != nil {
+		return result, err
+	}
+
+	if err = json.Unmarshal(respBody, &result); err != nil {
+		return result, err
+	}
+
+	return result, nil
```
Comment (anderson-comvex) [inline auth.go]:
```suggestion
	err = json.Unmarshal(respBody, &result)

	return result, err
```

**Review by anderson-comvex — CHANGES_REQUESTED** [review]

File: error.go (hunk 21 above)
Comment (ghost) [reply error.go:33]: sure! basically it just checks the response body if it's empty. If it is, we assume that it's an HTTP error from the server

File: examples/example.go [hunk 24]
```diff
@@ -0,0 +1,179 @@
...
+			},
+			BaseURL: "{{test_base_url}}",
+			Cache:   cache.New(5*time.Minute, 10*time.Minute),
+		},
+	}
+
+	res, err := uipath.GetOAuthToken(e.Client)
+	if err != nil {
+		fmt.Println(res, err)
+	}
```
Comment (cliff-comvex) [inline examples/example.go]: This is just an example but we don't need this anymore every time we want to execute a handler right? Maybe we can remove this here then?

**Review by cliff-comvex — CHANGES_REQUESTED** [review]

**Review by cliff-comvex — COMMENTED** [review]
Thanks for the `example_test.go`, this is very useful when creating feature tests in Lead-Api microservice.

File: examples/example.go [hunk 25]
```diff
@@ -0,0 +1,174 @@
...
+
+	e := &Examples{
+		Client: &uipath.Client{
+			HttpClient: &http.Client{Transport: tr},
+			Credentials: uipath.Credentials{
+				ClientID:   "{{test_client_Id}}",
+				UserKey:    "{{test_user_key}}",
+				TenantName: "{{test_tenant_name}}",
+			},
+			BaseURL: "{{test_base_url}}",
```
Comment (cliff-comvex) [inline examples/example.go]: What is this `"{{test_base_url}}" here? Can we add a comment?

**Review by cliff-comvex — CHANGES_REQUESTED** [review]

**Review by cliff-comvex — APPROVED** [review]

File: asset.go [hunk 26]
```diff
@@ -0,0 +1,152 @@
...
+	}
+
+	err = json.Unmarshal(resp, &result)
+
+	return result, err
+}
+
+// Update updates an asset
+func (a *AssetHandler) Update(asset Asset) (Asset, error) {
+
```
Comment (anderson-comvex) [inline asset.go:135]:
```suggestion
```suggestion
```
```

File: asset.go [hunk 27]
```diff
@@ -0,0 +1,152 @@
...
+	url := fmt.Sprintf("%s%s(%d)", a.Client.BaseURL, AssetEndpoint, asset.ID)
```
Comment (anderson-comvex) [inline asset.go]: We don't need this extra line:

```suggestion
	url := fmt.Sprintf("%s%s(%d)", a.Client.BaseURL, AssetEndpoint, asset.ID)
```

**Review by anderson-comvex — APPROVED** [review]

**Review by CJButton — APPROVED** [review]

**Pull request merged**